### PR TITLE
Add change password feature

### PR DESCRIPTION
**Pull request by Mrtenz**

### CODE REVIEW

**Comment by Mrtenz** [comment]
[sc-8612]

**Comment by Mrtenz** [comment]
Currently blocked by design.

File: packages/quill/src/app/screens/ChangePassword.tsx [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+};
+
+export const ChangePassword: FunctionComponent = () => {
+  const dispatch = useDispatch();
+  const loggingIn = useSelector(getLoggingIn);
+  const error = useSelector((state) => state.auth.error);
+
+  const form = useForm(
+    initialValues,
+    yupValidator(PASSWORD_SCHEMA.concat(SCHEMA), {
```
Comment (FrederikBolding) [inline packages/quill/src/app/screens/ChangePassword.tsx:38]: Can we move everything to the definition of `SCHEMA`?

File: packages/quill/src/crypto/secrets.ts [hunk 2]
```diff
@@ -100,3 +103,19 @@ export const getSalt = async (): Promise<Buffer> => {
 
   return newSalt;
 };
+
+export const deleteSalt = async () => {
+  await keytar.deletePassword(KEYTAR_SERVICE, KEYTAR_SALT_NAME);
```
Comment (FrederikBolding) [inline packages/quill/src/crypto/secrets.ts:108]: Could this just return the promise?

File: packages/quill/src/crypto/auth.sagas.ts [hunk 3]
```diff
@@ -67,6 +76,37 @@ export function* createPasswordWorker({ payload }: PayloadAction<string>) {
...
+export function* changePasswordWorker({ payload }: ReturnType<typeof changePassword>) {
+  try {
+    const isEqual: boolean = yield call(comparePassword, payload.currentPassword);
+    if (!isEqual) {
+      yield put(changePasswordFailed(translateRaw('CURRENT_PASSWORD_NOT_EQUAL')));
+      return;
+    }
+
+    const credentials: { account: string }[] = yield call(keytar.findCredentials, KEYTAR_SERVICE);
+    const privateKeys: [account: TUuid, privateKey: string][] = yield all(
```
Comment (FrederikBolding) [inline packages/quill/src/crypto/auth.sagas.ts:88]: Does this not need a `call` somewhere or is it okay to pass Promises directly to `all`?

File: packages/quill/src/crypto/secrets.ts [hunk 4]
```diff
@@ -24,7 +24,10 @@ export const init = async (password: string) => {
   await setEncryptionKey(key);
 };
 
-const savePrivateKey = (uuid: TUuid | typeof KEYTAR_SETTINGS_KEY_NAME, privateKey: string) => {
+export const savePrivateKey = (
```
Comment (FrederikBolding) [inline packages/quill/src/crypto/secrets.ts:27]: Why the export?

**Review by FrederikBolding — CHANGES_REQUESTED** [review]

File: packages/quill/src/crypto/auth.sagas.ts (hunk 3 above)
Comment (Mrtenz) [reply packages/quill/src/crypto/auth.sagas.ts:88]: AFAIK it's just the `redux-saga` way of doing `Promise.all` so I don't think this needs a `call`.

File: packages/quill/src/crypto/secrets.ts (hunk 4 above)
Comment (Mrtenz) [reply packages/quill/src/crypto/secrets.ts:27]: It's used in `changePasswordWorker`.

File: packages/quill/src/crypto/auth.sagas.ts (hunk 3 above)
Comment (FrederikBolding) [reply packages/quill/src/crypto/auth.sagas.ts:88]: Can't find anything about that in the docs 🤔 
https://redux-saga.js.org/docs/api/#alleffects---parallel-effects

File: packages/quill/src/app/screens/ChangePassword.tsx [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+const initialValues = {
+  currentPassword: '',
+  password: '',
+  passwordConfirmation: ''
+};
+
+export const ChangePassword: FunctionComponent = () => {
+  const dispatch = useDispatch();
+  const loggingIn = useSelector(getLoggingIn);
+  const error = useSelector((state) => state.auth.error);
```
Comment (FrederikBolding) [inline packages/quill/src/app/screens/ChangePassword.tsx:36]: Is there a reason this is not a selector? 🤔

File: packages/quill/src/crypto/auth.sagas.test.ts [hunk 6]
```diff
@@ -96,6 +105,55 @@ describe('createPasswordWorker', () => {
...
+    (keytar.findCredentials as jest.MockedFunction<
+      typeof keytar.findCredentials
+    >).mockImplementationOnce(async () => [
+      {
+        account: fAccount.uuid,
+        password: 'foo'
+      }
+    ]);
+
+    (safeGetPrivateKey as jest.MockedFunction<typeof safeGetPrivateKey>).mockImplementation(
```
Comment (FrederikBolding) [inline packages/quill/src/crypto/auth.sagas.test.ts:119]: Can we add a test for when this is not the case?

File: packages/quill/src/crypto/secrets.ts [hunk 7]
```diff
@@ -38,6 +41,18 @@ export const getPrivateKey = async (uuid: TUuid | typeof KEYTAR_SETTINGS_KEY_NAM
   return null;
 };
 
+/**
+ * A safe version of the function above, which does not throw an error if the private key cannot be
+ * decrypted.
+ */
+export const safeGetPrivateKey = async (uuid: TUuid | typeof KEYTAR_SETTINGS_KEY_NAME) => {
```
Comment (FrederikBolding) [inline packages/quill/src/crypto/secrets.ts:48]: Test please!

File: packages/quill/src/crypto/secrets.ts [hunk 8]
```diff
@@ -38,6 +41,18 @@ export const getPrivateKey = async (uuid: TUuid | typeof KEYTAR_SETTINGS_KEY_NAM
...
+  try {
+    return getPrivateKey(uuid);
+  } catch {
+    return null;
```
Comment (FrederikBolding) [inline packages/quill/src/crypto/secrets.ts:52]: Does try catch work when you directly return the promise? 🤔

File: packages/quill/src/crypto/auth.sagas.ts [hunk 9]
```diff
@@ -67,6 +76,49 @@ export function* createPasswordWorker({ payload }: PayloadAction<string>) {
   yield put(push(ROUTE_PATHS.SETUP_ACCOUNT));
 }
 
+export function* getAccountPrivateKey(account: TUuid) {
```
Comment (FrederikBolding) [inline packages/quill/src/crypto/auth.sagas.ts:79]: Does this need to be exported?

File: packages/quill/src/app/screens/ChangePassword.tsx (hunk 5 above)
Comment (Mrtenz) [reply packages/quill/src/app/screens/ChangePassword.tsx:36]: Nope

File: packages/quill/src/crypto/auth.sagas.ts (hunk 9 above)
Comment (Mrtenz) [reply packages/quill/src/crypto/auth.sagas.ts:79]: I intended to test it, but forgot. So yes

**Pull request merged**